code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/selector/SelectorUtils.java

```diff
@@ -0,0 +1,137 @@
+package io.cattle.platform.servicediscovery.selector;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+public class SelectorUtils {
+
+    private static class SelectorConstraintDefinition {
+        private enum SelectorOp {
+            NEQ("!=", false),
+            EQ("=", false),
+            NOTIN("notin", true),
+            IN("in", true),
+            NOOP("", false);
+
+            String selectorSymbol;
+            boolean isList;
+
+            private SelectorOp(String envSymbol, boolean isList) {
+                this.selectorSymbol = envSymbol;
+                this.isList = isList;
+            }
+
+            public String getSelectorSymbol() {
+                return selectorSymbol;
+            }
+        }
+
+        private SelectorOp op;
+        private String key;
+        private List<String> value = new ArrayList<>();
+
+        private SelectorConstraintDefinition(String selector) {
+            for (SelectorOp op : SelectorOp.values()) {
+                if (op == SelectorOp.NOOP) {
+                    continue;
+                }
+                String[] exp = selector.split(op.getSelectorSymbol());
+                if (exp.length == 2) {
+                    this.key = exp[0].trim();
+                    if (op.isList) {
+                        List<String> values = Arrays.asList(exp[1].split(","));
+                        for (String value : values) {
+                            this.value.add(value.trim());
+                        }
+                    } else {
+                        this.value.add(exp[1].trim());
+                    }
+                    this.op = op;
+                    break;
+                }
+            }
+            if (this.op == null) {
+                this.op = SelectorOp.NOOP;
+                this.key = selector.trim();
+            }
+        }
+
+        private SelectorOp getOp() {
+            return op;
+        }
+
+        private String getKey() {
+            return key;
+        }
+
+        private List<String> getValue() {
+            return value;
+        }
+    }
+
+    public static boolean isSelectorMatch(String selector, Map<String, String> labels) {
+        List<SelectorConstraintDefinition> constraints = getSelectorConstraints(selector);
```

Comment:
Is it worth protecting against a constraints.length() == 0 situation or is that impossible/highly unlikely?
Theoretically, if constraints.length() == 0, you should just return false?